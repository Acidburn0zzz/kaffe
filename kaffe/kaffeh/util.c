/*
 * util.c
 * Extra routine need for support.
 *
 * Copyright (c) 1996, 1997
 *	Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution 
 * of this file. 
 */

#include "config.h"
#include "config-std.h"
#include "config-mem.h"
#include "jtypes.h"
#include "constants.h"

Utf8Const*
makeUtf8Const(char* s, int len)
{
	Utf8Const* m;

	m = (Utf8Const*)malloc(sizeof(Utf8Const) + len + 1);
	memcpy(m->data, s, len);
	m->data[len] = 0;
	m->length = len;
	return (m);
}

#if !INTERN_UTF8CONSTS

/*
 * Count number of unicode characters in UtF8 string.
 */
int
strLengthUtf8(char* str, int len)
{
	register unsigned char* ptr;
	register unsigned char* limit;
	int str_length;

	ptr = (unsigned char*) str;
	limit = ptr + len;
	str_length = 0;
	for (; ptr < limit; str_length++) {
		if (UTF8_GET (ptr, limit) < 0) {
			return (-1);
		}
	}
	return (str_length);
}

/* Calculate a hash value for a string encoded in Utf8 format.
 * This returns the same hash value as specified or java.lang.String.hashCode.
 */
int32
hashUtf8String (char* str, int len)
{
	int str_length;
	register unsigned char* ptr;
	register unsigned char* limit;
	int32 hash;

	str_length = strLengthUtf8 (str, len);
	ptr = (unsigned char*) str;
	limit = ptr + len;
	hash = 0;

	if (str_length <= 15) {
		for (; ptr < limit;) {
			int ch = UTF8_GET(ptr, limit);
			hash = (37 * hash) + ch;
		}
	}
	else {
		int i = 0;
		for (; i < str_length;  i++) {
			int ch = UTF8_GET(ptr, limit);
			if ((i & 7) == 0) {
				hash = (39 * hash) + ch;
			}
		}
	}
	return (hash);
}

#endif
