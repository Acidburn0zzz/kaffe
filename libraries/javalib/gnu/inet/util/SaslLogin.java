/*
 * SaslLogin.java
 * Copyright (C) 2004 The Free Software Foundation
 * 
 * This file is part of GNU inetlib, a library.
 * 
 * GNU inetlib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * GNU inetlib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * As a special exception, if you link this library with other files to
 * produce an executable, this library does not by itself cause the
 * resulting executable to be covered by the GNU General Public License.
 * This exception does not however invalidate any other reasons why the
 * executable file might be covered by the GNU General Public License.
 */

package gnu.inet.util;

import java.io.UnsupportedEncodingException;
import javax.security.sasl.SaslClient;
import javax.security.sasl.SaslException;

/**
 * SASL mechanism for LOGIN.
 *
 * @author <a href='mailto:dog@gnu.org'>Chris Burdess</a>
 */
public class SaslLogin
implements SaslClient
{

  private static final int STATE_USERNAME = 0;
  private static final int STATE_PASSWORD = 1;
  private static final int STATE_COMPLETE = 2;

  private String username;
  private String password;
  private int state;

  public SaslLogin (String username, String password)
  {
    this.username = username;
    this.password = password;
    state = STATE_USERNAME;
  }

  public String getMechanismName ()
  {
    return "LOGIN";
  }

  public boolean hasInitialResponse ()
  {
    return false;
  }

  public byte[] evaluateChallenge (byte[] challenge)
    throws SaslException
  {
    try
      {
        switch (state)
          {
          case STATE_USERNAME:
            state = STATE_PASSWORD;
            return username.getBytes ("UTF-8");
          case STATE_PASSWORD:
            state = STATE_COMPLETE;
            return password.getBytes ("UTF-8");
          default:
            return new byte[0];
          }
      }
    catch (UnsupportedEncodingException e)
      {
        String msg = "The UTF-8 character set is not supported by the VM";
        throw new SaslException (msg, e);
      }
  }

  public boolean isComplete ()
  {
    return (state == STATE_COMPLETE);
  }

  public byte[] unwrap (byte[] incoming, int off, int len)
    throws SaslException
  {
    byte[] ret = new byte[len - off];
    System.arraycopy (incoming, off, ret, 0, len);
    return ret;
  }

  public byte[] wrap (byte[] outgoing, int off, int len)
    throws SaslException
  {
    byte[] ret = new byte[len - off];
    System.arraycopy (outgoing, off, ret, 0, len);
    return ret;
  }

  public Object getNegotiatedProperty (String name)
  {
    return null;
  }

  public void dispose ()
  {
  }
  
}
