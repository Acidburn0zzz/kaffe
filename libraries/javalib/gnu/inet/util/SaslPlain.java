/*
 * SaslPlain.java
 * Copyright (C) 2004 The Free Software Foundation
 * 
 * This file is part of GNU inetlib, a library.
 * 
 * GNU inetlib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * GNU inetlib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * As a special exception, if you link this library with other files to
 * produce an executable, this library does not by itself cause the
 * resulting executable to be covered by the GNU General Public License.
 * This exception does not however invalidate any other reasons why the
 * executable file might be covered by the GNU General Public License.
 */

package gnu.inet.util;

import java.io.UnsupportedEncodingException;
import javax.security.sasl.SaslClient;
import javax.security.sasl.SaslException;

/**
 * SASL mechanism for PLAIN.
 *
 * @author <a href='mailto:dog@gnu.org'>Chris Burdess</a>
 */
public class SaslPlain
implements SaslClient
{

  private String username;
  private String password;
  private boolean complete;

  public SaslPlain (String username, String password)
  {
    this.username = username;
    this.password = password;
  }

  public String getMechanismName ()
  {
    return "PLAIN";
  }

  public boolean hasInitialResponse ()
  {
    return true;
  }

  public byte[] evaluateChallenge (byte[] challenge)
    throws SaslException
  {
    try
      {
        byte[] a = username.getBytes ("UTF-8");
        byte[] b = password.getBytes ("UTF-8");
        byte[] c = new byte[(a.length * 2) + b.length + 2];
        System.arraycopy (a, 0, c, 0, a.length);
        System.arraycopy (a, 0, c, a.length + 1, a.length);
        System.arraycopy (b, 0, c, (a.length * 2) + 2, b.length);
        complete = true;
        return c;
      }
    catch (UnsupportedEncodingException e)
      {
        String msg = "Username or password contains illegal UTF-8";
        throw new SaslException (msg, e);
      }
  }

  public boolean isComplete ()
  {
    return complete;
  }

  public byte[] unwrap (byte[] incoming, int off, int len)
    throws SaslException
  {
    byte[] ret = new byte[len - off];
    System.arraycopy (incoming, off, ret, 0, len);
    return ret;
  }

  public byte[] wrap (byte[] outgoing, int off, int len)
    throws SaslException
  {
    byte[] ret = new byte[len - off];
    System.arraycopy (outgoing, off, ret, 0, len);
    return ret;
  }

  public Object getNegotiatedProperty (String name)
  {
    return null;
  }

  public void dispose ()
  {
  }
  
}
