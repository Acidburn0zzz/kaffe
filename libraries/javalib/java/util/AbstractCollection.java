/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

package java.util;

public abstract class AbstractCollection implements Collection {

public AbstractCollection() {
}

public abstract Iterator iterator();

public abstract int size();

public boolean isEmpty() {
  return (size() == 0);
}

public boolean contains(Object o) {
  Iterator it = iterator();
  while (it.hasNext()) {
    if (it.next() == o) {
      return (true);
    }
  }
  return (false);
}

public Object[] toArray() {
  Object[] objs = new Object[size()];
  Iterator it = iterator();
  for (int i = 0; i < objs.length; i++) {
    objs[i] = it.next();
  }
  return (objs);
}

public Object[] toArray(Object[] a) {
  throw new kaffe.util.NotImplemented();
}

public boolean add(Object o) {
  throw new UnsupportedOperationException();
}

public boolean remove(Object o) {
  Iterator it = iterator();
  while (it.hasNext()) {
    if (it.next() == o) {
      it.remove();
      return (true);
    }
  }
  return (false);
}

public boolean containsAll(Collection c) {
  Iterator it = c.iterator();
  while (it.hasNext()) {
    if (!contains(it.next())) {
      return (false);
    }
  }
  return (true);
}

public boolean addAll(Collection c) {
  Iterator it = c.iterator();
  while (it.hasNext()) {
    add(it.next());
  }
  return (true);
}

public boolean removeAll(Collection c) {
  Iterator it = c.iterator();
  boolean status = false;
  while (it.hasNext()) {
    if (remove(it.next())) {
      status = true;
    }
  }
  return (status);
}

public boolean retainAll(Collection c) {
  Iterator it = iterator();
  boolean status = false;
  while (it.hasNext()) {
    if (!c.contains(it.next())) {
      it.remove();
      status = true;
    }
  }
  return (status);
}

public void clear() {
  Iterator it = iterator();
  while (it.hasNext()) {
    it.remove();
  }
}

public String toString() {
  StringBuffer buf = new StringBuffer();
  Iterator it = iterator();
  buf.append("[");
  while (it.hasNext()) {
    Object o = it.next();
    buf.append(String.valueOf(o));
    if (it.hasNext()) {
      buf.append(", ");
    }
  }
  buf.append("]");

  return (buf.toString());
}

}
