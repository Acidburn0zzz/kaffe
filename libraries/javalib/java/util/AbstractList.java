/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

package java.util;

public abstract class AbstractList
  extends AbstractCollection implements List {

protected int modCount;

protected AbstractList() {
}
    
public boolean add(Object o) {
  add(size(), o);
  return (true);
}

public abstract Object get(int index);

public Object set(int index, Object element) {
  throw new UnsupportedOperationException();
}

public void add(int index, Object element) {
  throw new UnsupportedOperationException();
}

public boolean remove(Object o) {
  throw new UnsupportedOperationException();
}

public Object remove(int index) {
  throw new UnsupportedOperationException();
}

public int indexOf(Object o) {
  ListIterator it = listIterator();
  int idx = 0;
  while (it.hasNext()) {
    if (it.next() == o) {
      return (idx);
    }
    idx++;
  }
  return (-1);
}

public int lastIndexOf(Object o) {
  ListIterator it = listIterator(size());
  int idx = size() - 1;
  while (it.hasPrevious()) {
    if (it.previous() == o) {
      return (idx);
    }
    idx--;
  }
  return (-1);
}

public void clear() {
    removeRange(0, size());
}

public boolean addAll(int index, Collection c) {
  Object[] objs = c.toArray();
  for (int i = 0; i < objs.length; i++) {
    add(index + i, objs[i]);
  }
  return (true);
}

public Iterator iterator() {
  throw new kaffe.util.NotImplemented();
}

public ListIterator listIterator() {
  return (listIterator(0));
}

public ListIterator listIterator(int index) {
  throw new kaffe.util.NotImplemented();
}

public List subList(int fromIndex, int toIndex) {
  throw new kaffe.util.NotImplemented();
}

public boolean equals(Object o) {
  if (o == this) {
    return (true);
  }
  if (!(o instanceof List)) {
    return (false);
  }
  List other = (List)o;
  if (size() != other.size()) {
    return (false);
  }
  ListIterator mlist = listIterator();
  ListIterator olist = other.listIterator();
  while (mlist.hasNext()) {
    final Object o1 = mlist.next();
    final Object o2 = olist.next();
    if (!(o1 == null ? o2 == null : o1.equals(o2))) {
      return (false);
    }
  }
  return (true);
}

public int hashCode() {
  int hashCode = 1;
  for (Iterator i = iterator(); i.hasNext(); ) {
    final Object obj = i.next();
    hashCode = 31*hashCode + (obj==null ? 0 : obj.hashCode());
  }
  return hashCode;
}

protected void removeRange(int fromIndex, int toIndex) {
  throw new kaffe.util.NotImplemented();
}
 
}
