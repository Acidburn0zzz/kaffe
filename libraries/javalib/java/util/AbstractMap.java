
/*
 * Java core library component.
 *
 * Copyright (c) 1999
 *	Archie L. Cobbs.  All rights reserved.
 * Copyright (c) 1999
 *	Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 *
 * Author: Archie L. Cobbs <archie@whistle.com>
 */

package java.util;

public abstract class AbstractMap implements Map {
	private Set keyset;
	private Collection valcol;

	protected AbstractMap() {
	}

	public int size() {
		return entrySet().size();
	}

	public boolean isEmpty() {
		return size() == 0;
	}

	public boolean containsValue(Object value) {
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			Object next = ((Map.Entry)i.next()).getValue();
			if (value == null ? next == null : value.equals(next)) {
				return true;
			}
		}
		return false;
	}

	public boolean containsKey(Object key) {
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			Object next = ((Map.Entry)i.next()).getKey();
			if (key == null ? next == null : key.equals(next)) {
				return true;
			}
		}
		return false;
	}

	public Object get(Object key) {
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			Map.Entry e = (Map.Entry)i.next();
			Object ekey = e.getKey();
			if (key == null ? ekey == null : key.equals(ekey)) {
				return e.getValue();
			}
		}
		return null;
	}

	public Object put(Object key, Object value) {
		throw new UnsupportedOperationException();
	}

	public Object remove(Object key) {
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			Map.Entry e = (Map.Entry)i.next();
			Object ekey = e.getKey();
			if (key == null ? ekey == null : key.equals(ekey)) {
				Object value = e.getValue();
				i.remove();
				return value;
			}
		}
		return null;
	}

	public void putAll(Map t) {
		for (Iterator i = t.entrySet().iterator(); i.hasNext(); ) {
			Map.Entry e = (Map.Entry)i.next();
			put(e.getKey(), e.getValue());
		}
	}

	public void clear() {
		entrySet().clear();
	}

	public Set keySet() {
		if (keyset != null) {
			return keyset;
		}
		keyset = new AbstractSet() {
			public int size() {
				return AbstractMap.this.size();
			}
			public boolean contains(Object o) {
				return containsKey(o);
			}
			public boolean remove(Object o) {
				int beforeSize = size();
				AbstractMap.this.remove(o);
				return size() != beforeSize;
			}
			public void clear() {
				AbstractMap.this.clear();
			}
			public Iterator iterator() {
				return new Iterator() {
					private Iterator i = entrySet().iterator();
					public boolean hasNext() {
						return i.hasNext();
					}
					public Object next() {
						return ((Map.Entry)i.next()).getKey();
					}
					public void remove() {
						i.remove();
					}
				};
			}
		};
		return keyset;
	}

	public Collection values() {
		if (keyset != null) {
			return keyset;
		}
		valcol = new AbstractCollection() {
			public int size() {
				return AbstractMap.this.size();
			}
			public void clear() {
				AbstractMap.this.clear();
			}
			public Iterator iterator() {
				return new Iterator() {
					private Iterator i = entrySet().iterator();
					public boolean hasNext() {
						return i.hasNext();
					}
					public Object next() {
						return ((Map.Entry)i.next()).getValue();
					}
					public void remove() {
						i.remove();
					}
				};
			}
		};
		return valcol;
	}

	public abstract Set entrySet();

	public boolean equals(Object o) {
		if (!(o instanceof Map)) {
			return false;
		}
		Map omap = (Map)o;
		if (omap.size() != size()) {
			return false;
		}
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			Map.Entry ent = (Map.Entry)i.next();
			if (!omap.containsKey(ent.getKey())) {
				return false;
			}
			Object oval = omap.get(ent.getKey());
			Object myval = ent.getValue();
			if (!(myval == null ?
			    oval == null : myval.equals(oval))) {
				return false;
			}
		}
		return true;
	}

	public int hashCode() {
		int hash = 0;
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			hash += i.next().hashCode();
		}
		return hash;
	}

	public String toString() {
		StringBuffer b = new StringBuffer(5 * size());
		b.append("{");
		boolean first = true;
		for (Iterator i = entrySet().iterator(); i.hasNext(); ) {
			Map.Entry ent = (Map.Entry)i.next();
			if (!first) {
				b.append(", ");
			}
			b.append(ent.getKey() + "=" + ent.getValue());
			first = false;
		}
		b.append("}");
		return b.toString();
	}

}

