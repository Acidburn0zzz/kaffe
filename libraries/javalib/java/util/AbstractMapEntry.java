
/*
 * Java core library component.
 *
 * Copyright (c) 1999
 *	Archie L. Cobbs.  All rights reserved.
 * Copyright (c) 1999
 *	Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 *
 * Author: Archie L. Cobbs <archie@whistle.com>
 */

package java.util;

abstract class AbstractMapEntry implements Map.Entry {
	protected final Object key;
	protected Object value;

	AbstractMapEntry(Object key, Object value) {
		this.key = key;
		this.value = value;
	}

	public Object getKey() {
		return key;
	}

	public Object getValue() {
		return value;
	}

	public Object setValue(Object newValue) {
		Object oldValue = value;
		changeValue(newValue);
		value = newValue;
		return oldValue;
	}

	// This should write the new value back through to the underlying Map
	protected abstract void changeValue(Object newValue);

	public boolean equals(Object o) {
		if (!(o instanceof Map.Entry)) {
			return false;
		}
		Map.Entry me = (Map.Entry)o;
		return (this.key == null ? me.getKey() == null
			: this.key.equals(me.getKey()))
		    && (this.value == null ? me.getValue() == null
			: this.value.equals(me.getValue()));
	}

	public int hashCode() {
		return (key == null ? 0 : key.hashCode())
		    ^ (value == null ? 0 : value.hashCode());
	}
}

