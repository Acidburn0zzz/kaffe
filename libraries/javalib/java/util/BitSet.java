/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

package java.util;

import java.io.Serializable;

public class BitSet
  implements Cloneable, Serializable
{
	int[] bits;
	int size;
	static int[] mask = {
	0x00000001,
	0x00000002,
	0x00000004,
	0x00000008,
	0x00000010,
	0x00000020,
	0x00000040,
	0x00000080,
	0x00000100,
	0x00000200,
	0x00000400,
	0x00000800,
	0x00001000,
	0x00002000,
	0x00004000,
	0x00008000,
	0x00010000,
	0x00020000,
	0x00040000,
	0x00080000,
	0x00100000,
	0x00200000,
	0x00400000,
	0x00800000,
	0x01000000,
	0x02000000,
	0x04000000,
	0x08000000,
	0x10000000,
	0x20000000,
	0x40000000,
	0x80000000
};
	static int[] invMask = {
	0xfffffffe,
	0xfffffffd,
	0xfffffffb,
	0xfffffff7,
	0xffffffef,
	0xffffffdf,
	0xffffffbf,
	0xffffff7f,
	0xfffffeff,
	0xfffffdff,
	0xfffffbff,
	0xfffff7ff,
	0xffffefff,
	0xffffdfff,
	0xffffbfff,
	0xffff7fff,
	0xfffeffff,
	0xfffdffff,
	0xfffbffff,
	0xfff7ffff,
	0xffefffff,
	0xffdfffff,
	0xffbfffff,
	0xff7fffff,
	0xfeffffff,
	0xfdffffff,
	0xfbffffff,
	0xf7ffffff,
	0xefffffff,
	0xdfffffff,
	0xbfffffff,
	0x7fffffff
};

public BitSet () {
	this( 32);
}

public BitSet ( int size ) {
	if ( size <= 32 ) {
		bits = new int[1];
		this.size = 32;
	}
	else {
		int n = (size+31) >> 5;
		bits = new int[n];
		this.size = n << 5;
	}
}

public void and ( BitSet other ) {
	int[]  ob = other.bits;
	int    i, n = ob.length;

	if ( n > bits.length )
		grow( n << 5);
		
	for ( i=0; i<n; i++ )
		bits[i] &= ob[i];
}

public void clear ( int bit ) {
	if ( bit < 32 ) {
		bits[0] &= invMask[bit];
	}
	else {
		if ( bit >= size )
			grow( bit+1);

		int n = bit >> 5;
		int i = bit - (n << 5);
	
		bits[n] &= invMask[i];
	}
}

public Object clone () {
	BitSet other = new BitSet( size);
	System.arraycopy( bits, 0, other.bits, 0, bits.length);
	return other;
}

public boolean equals ( Object obj ) {
	if ( obj instanceof BitSet ) {
		BitSet other = (BitSet) obj;

		int[]  ob = other.bits;
		int    i, n = ob.length;

		if ( n != bits.length )
			return false;
		
		if ( n == 1 ){
			return (bits[0] == ob[0]);
		}
		else {
			for ( i=0; i<n; i++ ){
				if ( bits[i] != ob[i] ) return false;
			}	
			return true;
		}
	}
	
	return false;
}

public boolean get ( int bit ) {
	if ( bit < 32 ) {
		return ((bits[0] & mask[bit]) != 0);
	}
	else {
		if ( bit >= size )
			return false;

		int n = bit >> 5;
		int i = bit - (n << 5);

		return ((bits[n] & mask[i]) != 0);
	}
}

void grow ( int newSize ) {
	int n = (newSize+31) >> 5;
	int[] newBits;
	
	// no need to play the stop and go game if this is continously incremented
	if ( n < (bits.length<<1) )
		n = bits.length<<1;
	
	newBits = new int[n];
	
	System.arraycopy( bits, 0, newBits, 0, bits.length);
	bits = newBits;
	size = n << 5;
}

public int hashCode () {
	// Hmm, is this specified ? We use a very simple scheme here
	int i, h, b, n = bits.length;
	
	if ( n == 1 ){
		return bits[0];
	}
	else {
		h = (bits[0] == 0) ? 19 : bits[0];
	
		for ( i=1; i<n; i++ ) {
			if ( (b = bits[i]) == 0 )
				b = i;
			h ^= (h * 65599) + b;
		}
		return h;
	}
}

public void or ( BitSet other ) {
	int[]  ob = other.bits;
	int    i, n = ob.length;

	if ( n > bits.length )
		grow( n << 5);
		
	for ( i=0; i<n; i++ )
		bits[i] |= ob[i];
}

public void set ( int bit ) {
	if ( bit < 32 ) {
		bits[0] |= mask[bit];
	}
	else {
		if ( bit >= size )
			grow( bit+1);

		int n = bit >> 5;
		int i = bit - (n << 5);

		bits[n] |= mask[i];
	}
}

public int size () {
	return size;
}

public String toString () {
	StringBuffer s = new StringBuffer( size); // rough guess
	int b, i, j, k, n = bits.length;
	
	s.append( '{');
	for ( i=0; i<n; i++ ) {
		if ( (b = bits[i]) != 0 ) {
			k = i << 5;
			for ( j=0; j<32; j++ ){
				if ( (b & mask[j]) != 0 ){
					s.append( k + j);
					s.append( ',');
				}
			}
		}
	}
	if ( (i = s.length()) > 1 )
		s.setCharAt( i - 1, '}');
	else
		s.append( '}');
	
	return s.toString();
}

public void xor ( BitSet other ) {
	int[]  ob = other.bits;
	int    i, n = ob.length;

	if ( n > bits.length )
		grow( n << 5);
		
	for ( i=0; i<n; i++ )
		bits[i] ^= ob[i];
}
}
