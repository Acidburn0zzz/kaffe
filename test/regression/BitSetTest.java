/**
 * A test for java.util.BitSet
 *
 * Written by Edouard Parmelan <Edouard.Parmelan@France.NCR.COM>
 */
import java.util.BitSet;

public class BitSetTest {

    private static void show(String name, BitSet b) {
	System.out.println(name + " = " + b + "\tsize " + b.size() + "\tlength " + b.length());
	//System.out.println(name + " = " + b + "\tsize " + b.size());
    }


    private static void equalTest(String name, BitSet a, BitSet b) {
	System.out.println(name + " " + a.equals(b) + " == " + b.equals(a));
    }

    private static void hashCodeTest() {
	BitSet a = new BitSet(64);
	BitSet b = new BitSet(128);

	a.set(1);
	a.set(13);

	b.set(1);
	b.set(13);

	show("a", a);
	show("b", b);
	equalTest("a == b", a, b);
	System.out.println("hashCode()"
		+ " a " + Integer.toString(a.hashCode(), 16)
		+ " b " + Integer.toString(b.hashCode(), 16)
		);
	System.out.println();
    }

    public static void main(String args[]) {

	hashCodeTest();

	BitSet a = new BitSet(0);
	BitSet b = new BitSet(0);
	BitSet c = new BitSet(0);
	BitSet d = null;

	a.set(0);
	a.set(1);
	a.set(66);
	show("a", a);
	show("a.clone()", (BitSet)a.clone());

	b.set(1);
	b.set(2);
	b.set(63);
	show("b", b);
	show("b.clone()", (BitSet)b.clone());
	System.out.println();

	show("c", c);
	equalTest("a == b", a, b);
	System.out.println();

	d = (BitSet)a.clone();
	d.or(b);
	show("a or b", d);
	show("clone()", (BitSet)d.clone());
	c = new BitSet(0); c.or(d); show("{} or d", c);
	equalTest("d == d.clone()", d, (BitSet)d.clone());
	System.out.println();

	d = (BitSet)a.clone();
	d.and(b);
	show("a and b", d);
	show("clone()", (BitSet)d.clone());
	c = new BitSet(0); c.or(d); show("{} or d", c);
	equalTest("d == d.clone()", d, (BitSet)d.clone());
	System.out.println();

	d = (BitSet)a.clone();
	d.xor(b);
	show("a xor b", d);
	show("clone()", (BitSet)d.clone());
	c = new BitSet(0); c.or(d); show("{} or d", c);
	equalTest("d == d.clone()", d, (BitSet)d.clone());
	System.out.println();

	d = (BitSet)a.clone();
	d.andNot(b);
	show("a andNot b = ", d);
	show("clone()", (BitSet)d.clone());
	c = new BitSet(0); c.or(d); show("{} or d", c);
	equalTest("d == d.clone()", d, (BitSet)d.clone());
	System.out.println();
    }
}
