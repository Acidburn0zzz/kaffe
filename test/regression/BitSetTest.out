a = {1, 13}	size 64	length 14
b = {1, 13}	size 128	length 14
a == b true == true
hashCode() a 24d0 b 24d0

a = {0, 1, 66}	size 128	length 67
a.clone() = {0, 1, 66}	size 128	length 67
b = {1, 2, 63}	size 64	length 64
b.clone() = {1, 2, 63}	size 64	length 64

c = {}	size 0	length 0
a == b false == false

a or b = {0, 1, 2, 63, 66}	size 128	length 67
clone() = {0, 1, 2, 63, 66}	size 128	length 67
{} or d = {0, 1, 2, 63, 66}	size 128	length 67
d == d.clone() true == true

a and b = {1}	size 128	length 2
clone() = {1}	size 128	length 2
{} or d = {1}	size 64	length 2
d == d.clone() true == true

a xor b = {0, 2, 63, 66}	size 128	length 67
clone() = {0, 2, 63, 66}	size 128	length 67
{} or d = {0, 2, 63, 66}	size 128	length 67
d == d.clone() true == true

a andNot b =  = {0, 66}	size 128	length 67
clone() = {0, 66}	size 128	length 67
{} or d = {0, 66}	size 128	length 67
d == d.clone() true == true

